packages/compiler/src/core/transformer.ts

```diff
@@ -0,0 +1,307 @@
+import {
+  Diagnostic,
+  NoTarget,
+  Transform,
+  TransformSet,
+  TransformSetRef,
+  TransformerDefinition,
+  TransformerResolvedDefinition,
+} from "./types.js";
+
+/**
+ * Minimal interface for transformer library data needed internally.
+ * The full TransformerLibraryInstance from types.ts includes module metadata
+ * that isn't needed for transformer registration.
+ */
+interface TransformerLibrary {
+  transformer: TransformerResolvedDefinition;
+}
+
+export interface Transformer {
+  /**
+   * Extend the current transform set with additional transforms.
+   * @param transformSet - The transform set configuration to apply
+   * @returns Diagnostics from processing the transform set
+   */
+  extendTransformSet(transformSet: TransformSet): Promise<readonly Diagnostic[]>;
+
+  /**
+   * Register a transformer library.
+   * @param name - The library name
+   * @param lib - Optional library instance (will be loaded if not provided)
+   */
+  registerTransformLibrary(name: string, lib?: TransformerLibrary): Promise<void>;
+
+  /**
+   * Execute all enabled transforms and create mutation engines.
+   * @returns The transformation result including diagnostics, program, and engines
+   */
+  transform(): TransformerResult;
+
+  /**
+   * Get the mutation engine for a specific transform.
+   * Useful for debugging and inspecting transform state.
+   * @param transformId - The fully qualified transform ID (e.g., "@library/transform-name")
+   * @returns The mutation engine if it exists, undefined otherwise
+   */
+  getEngine(transformId: string): MutationEngine<any> | undefined;
+}
+
+export interface TransformerStats {
+  runtime: {
+    /** List of transform IDs that were enabled */
+    enabledTransforms: readonly string[];
+    /** Total time for all transform operations in milliseconds */
+    total: number;
+    /** Time spent creating each engine, keyed by transform ID */
+    engineCreation: Record<string, number>;
+  };
+}
+export interface TransformerResult {
+  readonly diagnostics: readonly Diagnostic[];
+  readonly program: TransformedProgram;
+  readonly stats: TransformerStats;
+  readonly engines: ReadonlyMap<string, MutationEngine<any>>;
+}
+
+/** Resolve a transformer definition for a library. */
+export function resolveTransformerDefinition(
+  libName: string,
+  transformer: TransformerDefinition,
+): TransformerResolvedDefinition {
+  const transforms: Transform<string>[] = transformer.transforms.map((t) => {
+    return { ...t, id: `${libName}/${t.name}` };
+  });
+  if (
+    transformer.transforms.length === 0 ||
+    (transformer.transformSets && "all" in transformer.transformSets)
+  ) {
+    return {
+      transforms,
+      transformSets: transformer.transformSets ?? {},
+    };
+  } else {
+    // Auto-generate an 'all' transform set that enables all transforms
+    const allEnable: Record<TransformSetRef, boolean> = {};
+    for (const t of transforms) {
+      allEnable[t.id as TransformSetRef] = true;
+    }
```

Comment:
did this not work?
```suggestion
    const allEnable = Object.fromEntries(transforms.map((x) => [x.id, true]));
```
I think TypeScript should properly infer the type of `allEnable` if you declare `transforms` to be of type `Transform<TransformSetRef>[]`.